### PR TITLE
Make FlaxPreTrainedModel a Flax Module

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

WIP. Makes `FlaxPreTrainedModel` a `nn.Module` so Flax users can easily integrate it into other Flax networks or systems that expect Flax Modules. 

See for #22499 some discussion of the approach.

### CODE REVIEW

**Comment by cgarciae** [comment]
@sanchit-gandhi's previous comment:

"""


This looks super clean @cgarciae - really like this way of making the FlaxPreTrainedModel into an nn.Module through dataclasses.

Had some questsions about the PR (mainly for my understanding), but think the general design philosophy looks good here.

From my testing it all seems to work as expected - think though we should add some very visible warning messages though when a user passes _do_init=False to advise them to use the returned model as a Flax nn.Module (rather than falling-back on the __call__ method of FlaxPreTrainedModel). This is the only breaking change I see from this PR, but one that is unavoidable (since it's the exact thing we're trying to change).

=> perhaps as a start we first make a PR that triggers this warning (advising users that the functinality is going to change in N months / releases time), and then have this PR as a follow-up that makes the changes?

For Flax BERT though this approach gives equivalence between the FlaxPreTrainedModel and the nn.Module -> for models that do extra pre-processing we can just modify the __call__ to do all the data pre-processing?

"""

**Comment by cgarciae** [comment]
Thanks for the review @sanchit-gandhi !

> => perhaps as a start we first make a PR that triggers this warning (advising users that the functinality is going to change in N months / releases time), and then have this PR as a follow-up that makes the changes?

I was planning on entirely deleting the `params` argument from `__call__` 😅. If we want to make the change a bit more gradual then maybe we could do something like this:

```diff
- if self._do_init:
+ if self.scope is None:
```
This would condition on whether the module being called inside `apply` or not.

**Comment by cgarciae** [comment]
@sanchit-gandhi cleaned the PR a little, these are the minimal changes required.

File: src/transformers/modeling_flax_utils.py [hunk 1]
```diff
@@ -170,46 +172,57 @@ class FlaxPreTrainedModel(PushToHubMixin, FlaxGenerationMixin):
         - **main_input_name** (`str`) -- The name of the principal input to the model (often `input_ids` for NLP
           models, `pixel_values` for vision models and `input_values` for speech models).
     """
-    config_class = None
-    base_model_prefix = ""
-    main_input_name = "input_ids"
+    # class attributes
+    # Ideally we could just use ClassVars here to indicate the types, but src/transformers/utils/import_utils.py
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:176]: Interesting! Hadn't used ClassVars for type checking before in transformers. Perhaps for this PR we can go for the simpler `else` statement and fix the ClassVars issue in a follow-up PR? E.g. collapse it all down to:
```python
     config_class = None
     base_model_prefix = None
     main_input_name = "input_ids"
```

File: src/transformers/modeling_flax_utils.py [hunk 2]
```diff
@@ -170,46 +172,57 @@ class FlaxPreTrainedModel(PushToHubMixin, FlaxGenerationMixin):
...
 
-        if _do_init:
+        if self._do_init:
             # randomly initialized parameters
-            random_params = self.init_weights(self.key, input_shape)
+            random_params = self.init_weights(self.key, self.input_shape)
             params_shape_tree = jax.eval_shape(lambda params: params, random_params)
         else:
-            init_fn = partial(self.init_weights, input_shape=input_shape)
+            init_fn = partial(self.init_weights, input_shape=self.input_shape)
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:225]: Nice! Thanks for iterating on this logic, think it's simpler to understand like this

File: src/transformers/modeling_flax_utils.py [hunk 3]
```diff
@@ -224,9 +237,22 @@ def __init__(
...
 
         # initialize the parameters
-        if _do_init:
+        if self._do_init:
             self.params = random_params
 
+        # must call super().__post_init__() to setup the flax module
+        super().__post_init__()
+
+    def __init_subclass__(cls, **kwargs):
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:246]: The magic method 🧙For the final PR we can add a doctoring explaining what it does / how it works.

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 4]
```diff
@@ -764,29 +766,46 @@ class FlaxBertPreTrainedModel(FlaxPreTrainedModel):
 
     config_class = BertConfig
     base_model_prefix = "bert"
-    module_class: nn.Module = None
+    if typing.TYPE_CHECKING:
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:769]: Again maybe cleaner to stick with `module_class = None` and address the ClassVar issue in a follow-up (/parallel) PR, WDYT?

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 5]
```diff
@@ -764,29 +766,46 @@ class FlaxBertPreTrainedModel(FlaxPreTrainedModel):
...
+        _config: BertConfig,
+        _module: Any = dataclasses.MISSING,  # unused, defined to comply with dataclass signature
         input_shape: Tuple = (1, 1),
         seed: int = 0,
         dtype: jnp.dtype = jnp.float32,
         _do_init: bool = True,
         gradient_checkpointing: bool = False,
+        *,
+        parent=None,
+        name=None,
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:785]: Are these two args required? Do we need them if `FlaxPreTrainedModel` is a nn.Module?

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 6]
```diff
@@ -890,63 +909,115 @@ def __call__(
...
             head_mask = jnp.ones((self.config.num_hidden_layers, self.config.num_attention_heads))
 
-        # Handle any PRNG if needed
-        rngs = {}
-        if dropout_rng is not None:
-            rngs["dropout"] = dropout_rng
+        if self.scope is None:
+            if not self._do_init and params is None:
+                raise ValueError(
+                    f"{self.__class__.__name__} was initialized with `_do_init=False` but `params` was provided."
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:915]:
```suggestion
                    f"{self.__class__.__name__} was initialized with `_do_init=False` but `params` were not provided."
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 7]
```diff
@@ -890,63 +909,115 @@ def __call__(
...
-        inputs = {"params": params or self.params}
+            # Handle any PRNG if needed
+            rngs = {}
+
+            if dropout_rng is not None:
+                rngs["dropout"] = dropout_rng
+
+            params = params or self.params
+
+            if "_module" in params:
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:926]: Would it be more user friendly to allow params to be passed as a dict without the extra `"_module"` nesting, e.g.
```diff
- {"params": {"_module": params}
+ {"params": params}
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 8]
```diff
@@ -890,63 +909,115 @@ def __call__(
         if head_mask is None:
             head_mask = jnp.ones((self.config.num_hidden_layers, self.config.num_attention_heads))
 
-        # Handle any PRNG if needed
-        rngs = {}
-        if dropout_rng is not None:
-            rngs["dropout"] = dropout_rng
+        if self.scope is None:
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:912]: So `self.scope` becomes our overarching way of telling whether the params should/shouldn't be passed? Is this better than `self._do_init`?

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 9]
```diff
@@ -890,63 +909,115 @@ def __call__(
...
+                    token_type_ids=jnp.array(token_type_ids, dtype="i4"),
+                    position_ids=jnp.array(position_ids, dtype="i4"),
+                    head_mask=jnp.array(head_mask, dtype="i4"),
+                    deterministic=not train,
+                    output_attentions=output_attentions,
+                    output_hidden_states=output_hidden_states,
+                    return_dict=return_dict,
+                )
+
+            # remove the cache from the output, the user should retrieve it from the cache collection
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:1018]: Cool! So the behaviour of the cache is unaffected if `_do_init=True` (i.e. stateful), but now needs to be handled in the cache collection if `_do_init=False` (i.e. stateless)

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for iterating @cgarciae! The `FlaxPreTrainedModel` class looks great. Left a few questions below about the `__call__` method for `FlaxBertPreTrainedModel`. In general, I think it's okay to have more complexity in the modelling code to handle the mix of stateful/stateless behaviour. Partitioning it with the `if/else` logic as you have done I think is the cleanest way of doing this.

File: src/transformers/models/bert/modeling_flax_bert.py (hunk 4 above)
Comment (cgarciae) [reply src/transformers/models/bert/modeling_flax_bert.py:769]: Sounds good, will remove all the `TYPE_CHECKING` usage

File: src/transformers/models/bert/modeling_flax_bert.py (hunk 8 above)
Comment (cgarciae) [reply src/transformers/models/bert/modeling_flax_bert.py:912]: It seems there are 3 ways this class would operate right now:
1. `_do_init=True`, the stafeful way.
2. `_do_init=False` but passing `model(..., params=params)`, the current "functional" way.
3. `_do_init=False` but using `mode.apply`, this results in `self.scope != None`.

Comment (sanchit-gandhi) [reply src/transformers/models/bert/modeling_flax_bert.py:912]: Right I see, thanks for clarifying!

**Comment by sanchit-gandhi** [comment]
Leaving closed in favour of #22866

**Pull request closed without merging**